### PR TITLE
Tests hard and recursion

### PR DESCRIPTION
**Pull request by AleksArnau**
- cleaned up hard.js, improved readability and changed variable names
- tests in hard.test.js successfull
- added test_files to .gitignore
- updated README.md

### CODE REVIEW

File: src/test_file.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+function fizzBuzz(start, end) {
```
Comment (peterthehan) [inline src/test_file.js]: lets remove this file if you were just using it to test stuff

File: test/test_file.test.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { describe, expect, test } from "@jest/globals";
```
Comment (peterthehan) [inline test/test_file.test.js]: also remove this file

File: test/hard.test.js [hunk 3]
```diff
@@ -1,5 +1,107 @@
...
+  {
+    input: ["milk", "egg", "butter", "butter", "shea butter", "egg"],
+    expected: ["egg", "butter"],
+  },
+  {
+    input: ["milk", "egg", "butter", "shea butter"],
+    expected: ["milk", "egg", "butter", "shea butter"],
+  },
+  { input: ["milk"], expected: ["milk"] },
+  { input: [], expected: undefined },
```
Comment (peterthehan) [inline test/hard.test.js]: lets actually return an empty list when nothing is given, fix the test and the implementation

File: test/hard.test.js [hunk 4]
```diff
@@ -1,5 +1,107 @@
...
+    input: ["milk"],
+    expected: "milk",
+  },
+  {
+    input: [],
+    expected: undefined,
+  },
+])("findMajority", ({ input, expected }) => {
+  test(`given ${input} should return ${expected}`, () => {
+    expect(findMajority(input)).toStrictEqual(expected);
```
Comment (peterthehan) [inline test/hard.test.js]: we can just use `.toBe` here instead of `.toStrictEqual`

File: src/hard.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      fizzBuzz.push(i);
+    }
+  }
+  return fizzBuzz;
+}
+
+function hasDuplicate(numbers) {
+  let numberTally = {};
+
+  for (let number = 0; number < numbers.length; number++) {
```
Comment (peterthehan) [inline src/hard.js]: the name `number` makes this kind of hard to read with the given input `numbers` when we're using number to mean the iteration index (`i` would be better)

even better still, we can just convert this into a for-of loop instead and skip that hassle

File: src/hard.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  let itemCounter = {};
+
+  for (let i = 0; i < items.length; i++) {
+    if (items[i] in itemCounter) {
+      itemCounter[items[i]]++;
+    } else {
+      itemCounter[items[i]] = 1;
+    }
```
Comment (peterthehan) [inline src/hard.js]: refactor: we can avoid using `else` by inverting the logic, a little more elegant imo

```suggestion
    if (!(items[i] in itemCounter)) {
      itemCounter[items[i]] = 0;
    }
    ++itemCounter[items[i]];
```

File: src/hard.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  let maxCount = 0;
+
+  for (const item in itemCounter) {
+    if (itemCounter[item] > maxCount) {
+      maxCount = itemCounter[item];
+    }
```
Comment (peterthehan) [inline src/hard.js]: refactor: we can do a nice trick here to make this a oneliner

```suggestion
    maxCount = Math.max(itemCounter[item], maxCount);
```

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,4 @@
 build
 node_modules
+src/test_files.js
+test/test_files.test.js
```
Comment (peterthehan) [inline .gitignore]: remove these two from the gitignore since you should delete these files from the PR

File: test/hard.test.js (hunk 4 above)
Comment (AleksArnau) [reply test/hard.test.js]: since an findMajority([]) now returns [], .toStrictEqual is required to pass the test

File: src/hard.js (hunk 6 above)
Comment (AleksArnau) [reply src/hard.js]: can't believe you're just allowed to put a ! in front of anything

Comment (AleksArnau) [reply src/hard.js:44]: I'll refactor all the times this logic shows up through medium.js and easy.js as well

File: src/hard.js (hunk 5 above)
Comment (AleksArnau) [reply src/hard.js]: made the for of loop, Do I still need to change `number` to `i`?

Comment (peterthehan) [reply src/hard.js]: no in this case with the for of loop, we are using the actual value and not an index so `number` is good

File: test/hard.test.js (hunk 4 above)
Comment (peterthehan) [reply test/hard.test.js]: I think you mixed up my previous comment, empty array for plurality, we can keep undefined for majority

Comment (peterthehan) [reply test/hard.test.js]: so we should use tobe

File: src/hard.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
+function fizzBuzz(start, end) {
+  let fizzBuzz = [];
```
Comment (peterthehan) [inline src/hard.js]: this array isn't being reassigned, we can use const

File: src/hard.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      fizzBuzz.push("buzz");
+    } else {
+      fizzBuzz.push(i);
+    }
+  }
+  return fizzBuzz;
+}
+
+function hasDuplicate(numbers) {
+  let numberTally = {};
```
Comment (peterthehan) [inline src/hard.js]: same here, const

File: src/hard.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+  return false;
+}
+
+function findPlurality(items) {
+  if (items.length === 0) {
+    return [];
+  }
+
+  let itemCounter = {};
```
Comment (peterthehan) [inline src/hard.js]: same here

File: src/hard.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+  for (let i = 0; i < items.length; i++) {
```
Comment (peterthehan) [inline src/hard.js]: nit: convert this to for-of

File: src/hard.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    ++itemCounter[items[i]];
+  }
+
+  let maxCount = 0;
+
+  for (const item in itemCounter) {
+    maxCount = Math.max(itemCounter[item], maxCount);
+  }
+
+  let pluralityList = [];
```
Comment (peterthehan) [inline src/hard.js]: same here

File: src/hard.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+  return pluralityList;
+}
+
+function findMajority(items) {
+  if (items.length === 0) {
+    return undefined;
+  }
+
+  let itemCounter = {};
```
Comment (peterthehan) [inline src/hard.js]: same here

File: src/hard.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  for (const item in itemCounter) {
+    if (itemCounter[item] === maxCount) {
+      pluralityList.push(item);
+    }
+  }
+
+  return pluralityList;
+}
+
+function findMajority(items) {
```
Comment (peterthehan) [inline src/hard.js:52]: I don't think this is using the most optimal solution, I would recheck your notes

**Review by peterthehan — APPROVED** [review]

**Pull request merged**